astrolabe/validator.py

```diff
@@ -106,194 +93,395 @@ def run_test_expecting_error(self, driver_workload):
             # run to error.
             pass
 
+        # The workload executor to permitted to exit prematurely when expecting
+        # an error, but it should still report stats via results.json
         try:
-            stats = subprocess.stop()
-        except PrematureExitError:
-            # OK for workload executor to exit prematurely when we expect
-            # it to fail with an error. We still need to return stats.
-            stats = subprocess.read_stats()
-        except TimeoutExpired:
-            self.fail("The workload executor did not terminate soon after "
-                      "receiving the termination signal.")
+            try:
+                stats = subprocess.stop()
+            except PrematureExitError:
+                stats = subprocess.read_stats()
+        except WorkloadExecutorError as exc:
+            self.fail("WorkloadExecutorError: %s" % exc)
```

Comment:
This is catching a potential exception from `subprocess.stop()` (re-thrown TimeoutExpired) and `subprocess.read_stats()`. In either case, we don't have stats to report so we're failing the test explicitly instead of using placeholder stats.